### DISCUSSION TITLE
关于自定义表单项的一些疑惑

### DISCUSSION THREAD

**Discussion by meowtec** [alibaba/formily#2494] [Q&A 2.x] [answered]
假设存在一个简单的表单组件，它接受了一个 value 并展示。用户点击后，异步 onChange 出一个新的随机 value。

```
function RandomBtn({ value, onChange }) {
  return (
    <button
      onClick={() => {
        setTimeout(() => onChange(Math.random()));
      }}
    >
      click: {value}
    </button>
  );
}
```

如果需要在 formily 表单中使用这个组件，需要怎么处理？（直接传给 createSchemaField 的 components 可以不？）

我写了一个在线例子，在这个例子里面，**有时候第一次点击是无效的**（即执行了 onChange 但是组件内容没有变化）：

https://codesandbox.io/s/vigilant-montalcini-hm707

是我的写法不对吗？我加了 connect 也是一样的结果。（在这个例子里面，加 connect 或者不加，有什么区别吗？）

另外，文档中说：

> 实现业务自定义组件主要是使用@formily/react 或@formily/vue中的 Hooks API 与 observer API
> 接入现成组件库的话，我们主要使用 connect/mapProps

这两种方式的区别又是什么？什么情况下需要使用第一种？有具体的例子吗？

**Comment by janryWang** [comment]
直接传给 createSchemaField 的 components 可以不？

这是标准用法，必须可以的

**Reply by meowtec** [reply]
那看看我的 playground 例子？
完整刷新整个页面，点击按钮，数据是有变化的。
点击 codesandbox 右侧的假浏览器的刷新按钮刷新子页面，这时首次点击按钮数据并没有刷新

**Reply by meowtec** [reply]
去掉 setTimeout 似乎没问题

**Reply by janryWang** [reply]
codesandbox有Bug

**Comment by janryWang** [comment]
connect只是用于桥接三方组件库，映射到字段模型中的属性才用的，具体看@formily/antd源码就知道了

**Comment by meowtec** [comment]
> codesandbox有Bug

不是 codesandbox 的问题，我在本地项目里面就发现了这个问题

**Reply by janryWang** [reply]
哦，你用了StrictMode，StrictMode下的确有问题，你直接单独提个Bug吧

**Reply by janryWang** [reply]
刚刚我这边排查了一下，formily内部字段已经重新渲染了，数据也更新了，但就是CustomBtn组件没重渲染，然后把setTimeout干掉或者加100毫秒延迟都没问题，具体我猜测，大概率是setTimeout跟fiber调度冲突了

**Reply by meowtec** [reply]
我试试去掉项目的 StrictMode

**Reply by meowtec** [reply]
我试了下 setTimeout 加 2000 也有问题

**Comment by janryWang** [comment, marked as answer]
这个是个Bug，已经修复 https://github.com/alibaba/formily/pull/2526
